File: 剑指offer/35.复杂链表的复制.py
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        mem = {}

        cur = head
        while cur:
            mem[cur] = Node(cur.val)
            cur = cur.next

        cur = head
        while cur:
            mem[cur].next = mem.get(cur.next)
            mem[cur].random = mem.get(cur.random)
            cur = cur.next

        return mem[head]

File: 剑指offer/35.复杂链表的复制.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node(int _val)
    {
        val = _val;
        next = NULL;
        random = NULL;
    }
};

class Solution {
public:
    Node* copyRandomList(Node* head)
    {
        unordered_map<Node*, Node*> map;
        Node* cur = head;

        while (cur) {
            Node* new_node = new Node(cur->val);
            map[cur] = new_node;
            cur = cur->next;
        }

        cur = head;
        while (cur) {
            map[cur]->next = map[cur->next];
            map[cur]->random = map[cur->random];
            cur = cur->next;
        }

        return map[head];
    }
};